tensorrt_llm/llmapi/utils.py

```diff
@@ -513,24 +515,45 @@ def get(self, timeout=None):
                 time.sleep(0.01)
 
 
-def set_sched_setaffinity(required_cores: int):
-    ''' Set the CPU affinity of the current process to the required number of
-    cores.
+def get_numa_aware_cpu_affinity(device_id):
+    '''Query NVML for NUMA-aware CPU affinity for the specified CUDA device.
 
-    Known issue: This may race with other processes that also set the affinity.
+    Args:
+        device_id: The CUDA device ID to query for optimal CPU affinity.
+
+    Returns:
+        List of CPU IDs representing the optimal CPU affinity mask for the device.
+
+    Raises:
+        pynvml.NVMLError: If NVML operations fail or device_id is invalid.
     '''
-    cpu_percentages = psutil.cpu_percent(percpu=True)
-    # sort the cores by usage
-    free_cores = sorted(range(len(cpu_percentages)),
-                        key=lambda i: cpu_percentages[i])
+    cpu_count = psutil.cpu_count()
+
+    # Get the number of bits per ulong
+    c_ulong_bits = ctypes.sizeof(ctypes.c_ulong) * 8
+
+    # Determine how large our cpu set array from NVML needs to be
+    cpu_set_size = math.ceil(cpu_count / c_ulong_bits)
+
+    # initialize NVML
```

Comment:
Is there a way to detect non-NUMA systems and early exit with the full list of CPUs?

Reply:
You can just check `os.path.isdir("/sys/devices/system/node/node1")` (thanks to Perplexity for the insight) and avoid needing to initialize NVML.